llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -234,76 +209,250 @@ def __init__(
+            remote_files: List of remote files in the save folder
+
+        Returns:
+            Dict of checkpoints that are complete and ready to be evaled
+        """
+        # Count the number of shards for each checkpoint group
+        remote_file_group_counts = Counter()
+        for f in remote_files:
+            interval_path = Path(f).parts[-2]
+            remote_file_group_counts[interval_path] += 1
+
+        # Check if all shards are present for each checkpoint group
+        checkpoints_to_eval = {}
+        for checkpoint, interval in checkpointer_checkpoints.items():
+            # eg {save_folder}/ep0-ba1/.
+            interval_path = Path(checkpoint).parts[-2]
+
+            # expecting one shard per gpu + 1 for metadata
+            expected_shard_count = dist.get_world_size() + 1
+            if remote_file_group_counts[interval_path] != expected_shard_count:
+                log.debug(
+                    f'Checkpoint {checkpoint} not fully uploaded (missing shards '
+                    +
+                    f'{remote_file_group_counts[interval_path]}/{expected_shard_count}), skipping'
+                )
+                continue
+
+            checkpoints_to_eval[interval_path] = interval
+
+        return checkpoints_to_eval
+
+    @staticmethod
+    def _get_ready_single_checkpoints(
+        checkpointer_checkpoints: Dict[str, Timestamp],
+        remote_checkpoints: List[str],
+    ) -> Dict[str, Timestamp]:
+        """Identify checkpoints ready to be evaled based on remote checkpoints.
+
+        This is much simpler than the sharded case, because there is only one file
+
+        Args:
+            checkpointer_checkpoints: All checkpoints from the checkpointer state
+            remote_checkpoints: List of remote checkpoints in the save folder
+
+        Returns:
+            Dict of checkpoints that are complete and ready to be evaled
+        """
+        unique_remote_checkpoints = set(remote_checkpoints)
+
+        checkpoints_to_eval = {}
+        for checkpoint, interval in checkpointer_checkpoints.items():
+            # eg {save_folder}/ep0-ba1-rank0.pt
+            interval_path = Path(checkpoint).parts[-1]
+
+            if checkpoint not in unique_remote_checkpoints:
+                log.debug(
+                    f'Checkpoint {checkpoint} not fully uploaded, skipping')
+                continue
+
+            checkpoints_to_eval[interval_path] = interval
+        return checkpoints_to_eval
+
+    def _get_checkpoints_and_launch_runs(self, state: State):
+        """Get the latest checkpoint from the training run.
+
+        Args:
+            state: The current state of the training run
+
+        Returns:
+            Returns checkpoints that have not been evaled
+        """
+        checkpointer = None
+        for callback in state.callbacks:
+            if isinstance(callback, CheckpointSaver):
+                checkpointer = callback
```

Comment:
just a little bit of proofing against strange uses...can we emit a warning if there are multiple checkpointsavers and say that this callback will be using the first one?